transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,439 @@
+    /// State of all open substreams.
+    substreams: HashMap<StreamId, SubstreamState>,
+    /// Pending substreams.
+    pending_substreams: VecDeque<Waker>,
+    /// Close waker.
+    close_waker: Option<Waker>,
+}
+
+/// State of a single substream.
+#[derive(Debug, Default)]
+struct SubstreamState {
+    /// Waker to wake if the substream becomes readable.
+    read_waker: Option<Waker>,
+    /// Waker to wake if the substream becomes writable.
+    write_waker: Option<Waker>,
+}
+
+impl QuicMuxer {
+    pub fn new(endpoint: ConnectionChannel, connection: QuinnConnection) -> Self {
+        Self {
+            inner: Mutex::new(QuicMuxerInner {
+                accept_incoming: false,
+                endpoint,
+                connection,
+                waker: None,
+                timer: None,
+                substreams: Default::default(),
+                pending_substreams: Default::default(),
+                close_waker: None,
+            }),
+        }
+    }
+
+    pub fn is_handshaking(&self) -> bool {
+        self.inner.lock().connection.is_handshaking()
+    }
+
+    pub fn peer_id(&self) -> Option<PeerId> {
+        let inner = self.inner.lock();
+        let session = inner.connection.crypto_session();
+        let identity = session.peer_identity()?;
+        let certs: Box<Vec<rustls::Certificate>> = identity.downcast().ok()?;
+        let cert = certs.get(0)?;
+        let p2p_cert = crate::tls::certificate::parse_certificate(&cert.0).ok()?;
+        Some(PeerId::from_public_key(&p2p_cert.extension.public_key))
+    }
+
+    pub fn local_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let ip = inner
+            .connection
+            .local_ip()
+            .unwrap_or(IpAddr::V4(Ipv4Addr::UNSPECIFIED));
+        let addr = SocketAddr::new(ip, inner.endpoint.port());
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub fn remote_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let addr = inner.connection.remote_address();
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub(crate) fn set_accept_incoming(&self, accept: bool) {
+        let mut inner = self.inner.lock();
+        inner.accept_incoming = accept;
+    }
+}
+
+impl StreamMuxer for QuicMuxer {
+    type Substream = StreamId;
+    type OutboundSubstream = ();
+    type Error = QuicMuxerError;
+
+    fn poll_event(
+        &self,
+        cx: &mut Context,
+    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
+        let mut inner = self.inner.lock();
+        let now = Instant::now();
+
+        while let Poll::Ready(event) = inner.endpoint.poll_channel_events(cx) {
+            inner.connection.handle_event(event);
+        }
+
+        let max_datagrams = inner.endpoint.max_datagrams();
+        while let Some(transmit) = inner.connection.poll_transmit(now, max_datagrams) {
+            inner.endpoint.send_transmit(transmit);
```

Comment:
Note I am not 100% sure this would adhere to the API requirements of `QuinnConnection`, namely to call `send_transmit` after a `handle_event`:

https://github.com/quinn-rs/quinn/blob/7b4b4c31e5878bdbce56ab2737dd8cfbafe1c4db/quinn-proto/src/connection/mod.rs#L89-L112

Not sure whether I am interpreting these requirements too strictly. 

In case you are interested in an alternative call sequence, see https://github.com/mxinden/rust-libp2p/commit/1230178cd660783a57b895d377621abaa66346e7 more specifically:

https://github.com/mxinden/rust-libp2p/commit/1230178cd660783a57b895d377621abaa66346e7#diff-d59efbc1f8dc6ed1521dd36e906216f4ff4ade561dca9b50beea1da354240cf7R144-R305